### PR TITLE
SE: Learn number constraints for +=, -=, &=, |=, ^=, *=, /= and %= 

### LINKED ISSUE #7144: SE: Add CompoundAssignmentOperation support for numeric values
* [x] Reset target symbols state on compound assignment
* [x] Learn NotNull for result of compound assignment (added after validation)
* [x] Add support for += and -=.
* [x] Add &=, |=, ^=, *=, /=, %=

Handling will be the same as in the BinaryOperation support #7141

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #7144

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.CompoundAssignment.cs [hunk 1]
```diff
@@ -63,13 +83,13 @@ public void CompoundAssignment_SelfUpdate()
...
-        validator.TagValue("Result").Should().HaveOnlyConstraints(ObjectConstraint.NotNull);
-        validator.TagValue("Value").Should().HaveOnlyConstraints(ObjectConstraint.NotNull);
+        validator.TagValue("Result").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(42));
+        validator.TagValue("Value").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(42));
     }
 
     [DataTestMethod]
-    [DataRow("+=")]
-    [DataRow("-=")]
+  // [DataRow("+=")]
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.CompoundAssignment.cs:91]: The plan is to merge on this https://github.com/SonarSource/sonar-dotnet/pull/7923
and then fix this test.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs [hunk 2]
```diff
@@ -18,18 +18,46 @@
...
-        var state = context.SetOperationValue(SymbolicValue.NotNull);
-        return assignment.Target.TrackedSymbol(state) is { } symbol
-            ? state.SetSymbolValue(symbol, SymbolicValue.NotNull)
-            : state;
+        var state = context.State;
+        var leftNumber = state[assignment.Target]?.Constraint<NumberConstraint>();
+        var rightNumber = state[assignment.Value]?.Constraint<NumberConstraint>();
+        if (leftNumber is { } && rightNumber is { } && Calculate(assignment.OperatorKind, leftNumber, rightNumber) is { } constraint)
+        {
+            state = state.SetSymbolConstraint(assignment.Target.TrackedSymbol(state), constraint);
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs:42]: I cannot find a case where `(assignment.Target.TrackedSymbol(state)` is null and has a number constrain, so I don't check.

Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs:42]: I don't think we can create one right now but I would definitely check for TrackedSymbol regardless for the code to be future-proof.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs [hunk 3]
```diff
@@ -18,18 +18,62 @@
...
+            return state.SetOperationConstraint(assignment, constraint);
+        }
+        else if (leftNumber is { }  && assignment.Target.TrackedSymbol(state) is { } targetSymbol  && state[assignment.Value]?.Constraint<ObjectConstraint>() is { Kind: ConstraintKind.NotNull })
+        {
+            return state.SetSymbolValue(targetSymbol, SymbolicValue.NotNull);
+        }
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs:48]: we can learn something for the operation if the target has a number constraint.

Cool, so this means we really don't need to check for the nullability of the target symbol right?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs [hunk 4]
```diff
@@ -18,18 +18,62 @@
...
+        return null;
     }
+
+    private static ProgramState ResetConstraintsAfterCompoundAssignment(ProgramState state, ICompoundAssignmentOperationWrapper assignment)
+    {
+        // We don't know what happens to the target after the compound assignment, unless it's a number.
+        if (assignment.Target.TrackedSymbol(state) is { } targetSymbol
+            && targetSymbol.GetSymbolType() is { } type
+            && !type.IsNullableValueType())
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs:57]: What I observed is that when the target of a nullable compound assignment is a nullable type and is null the result is always null.
So it's a good info to keep.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs]: You are setting a SymbolValue here, not an OperationConstraint. I still don't know what you are trying to achieve. Can you show me a code example?

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM

**Comment by mary-georgiou-sonarsource** [comment]
Converting to draft until I get feedback for the placement of the Arithmetics class.

**Pull request merged**